src/distrs/wilcox.jl

```diff
@@ -0,0 +1,152 @@
+#=
+The wilcox distribution is equivalent to the problem
+of finding the number of subsets with size nx
+of {1,2,...,nx,nx+1,...,nx+ny} summing to (U+sum(1:nx))
+relative to the total number of subsets with size nx.
+The empty subset is defined to sum to zero.
+This can be calculated using the recursion:
+either nx+ny is in the subset in which case we need to calculate
+the number of subsets with size nx-1 of {1,2,...,nx,nx+1,...,nx+ny-1}
+summing to (U+sum(1:nx)-nx-ny),
+or nx+ny is not in the subset in which case we need to calculate
+the number of subsets with size nx of {1,2,...,nx,nx+1,...,nx+ny-1}
+summing to (U+sum(1:nx)),
+This can be calculated bottom up using dynamic programming.
+
+The (i,j)'th element of DP in the k'th outer loop iteration represents:
+the number of subsets with size k of {1,2,...,k,k+1,...,k+j-1} summing to i+sum(1:k)-1.
+ =#
+
+@inline function wilcoxDP(nx, ny, U)
+    DP = zeros(Int, U + 1, ny + 1)
+    for j = 1:(ny+1)
+        DP[1, j] = 1
+    end
+    for k = 1:nx
+        for j = 2:(ny+1)
+            local_max = k * (j - 1)
+            for i = min(U + 1, local_max + 1):-1:max(j, 2)
+                # In this loop: min(U+1,local_max+1) >= i >= 2 AND i - j >= 0
+                DP[i, j] = DP[i-j+1, j] + DP[i, j-1]
+            end
+            for i = min(U + 1, local_max + 1, j - 1):-1:2
+                # In this loop: min(U+1,local_max+1) >= i >= 2 AND i - j < 0
```

Comment:
Given https://github.com/JuliaStats/StatsFuns.jl/pull/174#discussion_r2019610941, shouldn't we use
```suggestion
            i_max = k * (j - 1) + 1 + min(U - nx * j, 0)
            for i = i_max:-1:max(j, 2)
                # In this loop: i_max >= i >= 2 AND i - j >= 0
                DP[i, j] = DP[i-j+1, j] + DP[i, j-1]
            end
            for i = min(i_max, j - 1):-1:2
                # In this loop: i_max >= i >= 2 AND i - j < 0
```

Reply:
This doesn't quite get the indexing right, rolling back to the previous version
https://github.com/JuliaStats/StatsFuns.jl/actions/runs/14151899528

Reply:
I think I made an error in my reasoning, re-reading it I think condition 2 doesn't make sense: It mixes maximum sum of a set of size k with the minimum sum of a set of size nx-k.

Maybe there's some additional improvement possible, but to derive a bound here it's certainly safe to only use information about whether there's a subset of size `k` of `{1,...,k+j-1}` that sums to `i+sum(1:k)-1`, and disregard whether such a subset can be used as part of a subset of size nx that sums up to `U + nx*(nx+1)/2` (where in the cdf calculations we're even interested in all sums `nx*(nx+1)/2, 1+nx*(nx+1)/2, ..., U+nx*(nx+1)/2`).

Generally, we know that the sum `s` of a subset of size `k` of `{1,...,k+j-1}` satisfies

$$
\frac{k(k+1)}{2} \leq s \leq \frac{k(2j+k-1}{2}
$$

(sum of the `k` smallest and `k` largest values).
Thus we know that the number of subsets that sum to `i+sum(1:k)-1 = i+k(k+1)/2-1` is zero if

$$
i + \frac{k(k+1)}{2} - 1 < \frac{k(k+1)}{2} \lor i + \frac{k(k+1)}{2} - 1 > \frac{k(2j+k-1}{2},
$$

one if

$$
i + \frac{k(k+1)}{2} - 1 = \frac{k(k+1)}{2} \lor i + \frac{k(k+1)}{2} - 1 = \frac{k(2j+k-1}{2},
$$

and non-zero for other values of `i`.

Thus, in the `k`th outer loop
1. `DP[i,j] = 0` if `i < 1` or `i > k*(j-1) + 1`,
2. `DP[i,j] = 1` if `i = 1` or `i = k*(j-1) + 1`,
3. `DP[i,j] > 0` otherwise.

Thus we can restrict the inner loop to `2:(k*(j-1)+1)` (`i = 1` is handled in the separate loop above). Which is already done in the latest commit in this PR.

To reduce computing the max. value for `i` multiple times, I'd simplify it to

```suggestion
            i_max = min(U, k * (j - 1)) + 1
            for i = i_max:-1:max(j, 2)
                # In this loop: i_max >= i >= 2 AND i - j >= 0
                DP[i, j] = DP[i-j+1, j] + DP[i, j-1]
            end
            for i = min(i_max, j - 1):-1:2
                # In this loop: i_max >= i >= 2 AND i - j < 0
```